# Setup

In [28]:
import pandas as pd
import numpy as np

In [ ]:
## SAVE FEATURE STATISTICS FROM PREPROCESSING ##
X_MEAN = 21.234056
Y_MEAN = -16.93773
Z_MEAN = 24.948428

HR_MEAN = 81.085869 # bpm
WEIGHT_MEAN = 73.272422 # kg
HEIGHT_MEAN = 172.332927 # cm
GENDER_MEAN = 0.721742 # 0:female, 1:male -- represents class imbalance
AGE_MEAN = 30.236566 # yrs

EE_MEAN = 3481.625380 # cal/min (NOT kcal/min!!!!!)

X_STD = 35.144448
Y_STD = 37.25645
Z_STD = 23.991521

HR_STD = 29.852576
WEIGHT_STD = 11.258137
HEIGHT_STD = 6.570272
GENDER_STD = 0.448141
AGE_STD = 5.199545

EE_STD = 2752.199323 # cal/min (NOT kcal/min!!!!!)

# CNNFCN

## Run 1 LOOCV

### Best model

In [3]:
CNNFCN_run1_path = 'results/CNNFCN_run1.csv'
CNNFCN_run1_df = pd.read_csv(CNNFCN_run1_path)
CNNFCN_run1_df

avg_train_mse  avg_train_rmse  avg_val_mse  avg_val_rmse  avg_val_r2  \
0         0.263814        0.513629     0.613907      0.783522    0.370740   
1         0.259250        0.509166     0.666617      0.816466    0.382786   
2         0.263542        0.513363     0.942815      0.970986   -0.143303   
3         0.267724        0.517421     0.701133      0.837337    0.361930   
4         0.291004        0.539448     0.574444      0.757921    0.436996   
..             ...             ...          ...           ...         ...   
139       1.177141        1.084961     1.207299      1.098772   -0.095337   
140       0.438081        0.661877     0.498180      0.705819    0.519610   
141       0.386622        0.621789     0.625880      0.791126    0.298825   
142       0.402782        0.634651     0.537016      0.732814    0.504749   
143       0.323156        0.568468     0.669193      0.818042    0.316889   

     avg_epochs                                         model_name  \
0     40.636364  CNNFCNRegressor_conv1_fcn1_fcnw32_lr0.005_batc...   
1     48.454545  CNNFCNRegressor_conv1_fcn1_fcnw32_lr0.005_batc...   
2     41.454545  CNNFCNRegressor_conv1_fcn1_fcnw64_lr0.005_batc...   
3     46.000000  CNNFCNRegressor_conv1_fcn1_fcnw64_lr0.005_batc...   
4     57.363636  CNNFCNRegressor_conv1_fcn1_fcnw32_lr0.001_batc...   
..          ...                                                ...   
139   30.454545  CNNFCNRegressor_conv3_fcn3_fcnw64_lr0.05_batch...   
140   39.090909  CNNFCNRegressor_conv3_fcn3_fcnw32_lr0.01_batch...   
141   33.363636  CNNFCNRegressor_conv3_fcn3_fcnw32_lr0.01_batch...   
142   33.454545  CNNFCNRegressor_conv3_fcn3_fcnw64_lr0.01_batch...   
143   35.090909  CNNFCNRegressor_conv3_fcn3_fcnw64_lr0.01_batch...   

      timestamp  checkpoint_dir_name  done  ...  config/num_conv_layers  \
0    1732767114                  NaN  True  ...                       1   
1    1732767121                  NaN  True  ...                       1   
2    1732767119                  NaN  True  ...                       1   
3    1732767106                  NaN  True  ...                       1   
4    1732767222                  NaN  True  ...                       1   
..          ...                  ...   ...  ...                     ...   
139  1732775080                  NaN  True  ...                       3   
140  1732775333                  NaN  True  ...                       3   
141  1732775255                  NaN  True  ...                       3   
142  1732775355                  NaN  True  ...                       3   
143  1732775354                  NaN  True  ...                       3   

    config/num_fcn_layers  config/fcn_width  \
0                       1                32   
1                       1                32   
2                       1                64   
3                       1                64   
4                       1                32   
..                    ...               ...   
139                     3                64   
140                     3                32   
141                     3                32   
142                     3                64   
143                     3                64   

                                        config/log_dir  \
0    /Users/tyler/GitHub Repositories/UT Austin/202...   
1    /Users/tyler/GitHub Repositories/UT Austin/202...   
2    /Users/tyler/GitHub Repositories/UT Austin/202...   
3    /Users/tyler/GitHub Repositories/UT Austin/202...   
4    /Users/tyler/GitHub Repositories/UT Austin/202...   
..                                                 ...   
139  /Users/tyler/GitHub Repositories/UT Austin/202...   
140  /Users/tyler/GitHub Repositories/UT Austin/202...   
141  /Users/tyler/GitHub Repositories/UT Austin/202...   
142  /Users/tyler/GitHub Repositories/UT Austin/202...   
143  /Users/tyler/GitHub Repositories/UT Austin/202...   

                                      config/model_dir  \
0    /Users/tyler/

In [5]:
CNNFCN_run1_min_rmse = CNNFCN_run1_df[CNNFCN_run1_df['avg_val_rmse'] == CNNFCN_run1_df['avg_val_rmse'].min()]
CNNFCN_run1_min_rmse

avg_train_mse  avg_train_rmse  avg_val_mse  avg_val_rmse  avg_val_r2  \
140       0.438081        0.661877      0.49818      0.705819     0.51961   

     avg_epochs                                         model_name  \
140   39.090909  CNNFCNRegressor_conv3_fcn3_fcnw32_lr0.01_batch...   

      timestamp  checkpoint_dir_name  done  ...  config/num_conv_layers  \
140  1732775333                  NaN  True  ...                       3   

    config/num_fcn_layers  config/fcn_width  \
140                     3                32   

                                        config/log_dir  \
140  /Users/tyler/GitHub Repositories/UT Austin/202...   

                                      config/model_dir  \
140  /Users/tyler/GitHub Repositories/UT Austin/202...   

    config/early_stop_patience config/fold_early_stop  config/print  \
140                         20                     20         False   

     config/tune config/loocv  
140         True         True  

[1 rows x 34 columns]

In [8]:
CNNFCN_run1_min_rmse['model_name'].values[0]

'CNNFCNRegressor_conv3_fcn3_fcnw32_lr0.01_batch16_epochs100'

In [9]:
CNNFCN_run1_max_r2 = CNNFCN_run1_df[CNNFCN_run1_df['avg_val_r2'] == CNNFCN_run1_df['avg_val_r2'].max()]
CNNFCN_run1_max_r2

avg_train_mse  avg_train_rmse  avg_val_mse  avg_val_rmse  avg_val_r2  \
140       0.438081        0.661877      0.49818      0.705819     0.51961   

     avg_epochs                                         model_name  \
140   39.090909  CNNFCNRegressor_conv3_fcn3_fcnw32_lr0.01_batch...   

      timestamp  checkpoint_dir_name  done  ...  config/num_conv_layers  \
140  1732775333                  NaN  True  ...                       3   

    config/num_fcn_layers  config/fcn_width  \
140                     3                32   

                                        config/log_dir  \
140  /Users/tyler/GitHub Repositories/UT Austin/202...   

                                      config/model_dir  \
140  /Users/tyler/GitHub Repositories/UT Austin/202...   

    config/early_stop_patience config/fold_early_stop  config/print  \
140                         20                     20         False   

     config/tune config/loocv  
140         True         True  

[1 rows x 34 columns]

In [10]:
CNNFCN_run1_max_r2['model_name'].values[0]

'CNNFCNRegressor_conv3_fcn3_fcnw32_lr0.01_batch16_epochs100'

In [61]:
CNN_RMSE = CNNFCN_run1_min_rmse['avg_val_rmse'].values[0]
CNN_RMSE = (CNN_RMSE * EE_STD + EE_MEAN) / 1000
CNN_RMSE

5.424180022477394

In [62]:
CNN_RMSE / WEIGHT_MEAN

0.0740275791958589

CNN RMSE:

5.424 kcal/min

0.0740 kcal/kg/min

In [63]:
CNN_CV = CNN_RMSE / (EE_MEAN / 1000) # kcal/min
CNN_CV

1.5579447615577164

### All models

In [47]:
CNNFCN_run1_avg_rmse = CNNFCN_run1_df['avg_val_rmse'].mean()
CNNFCN_run1_avg_rmse

0.8770630396403608

In [49]:
(CNNFCN_run1_avg_rmse * EE_STD + EE_MEAN) / 1000

5.895477683926523

In [50]:
CNN_FCN_run1_std_rmse = CNNFCN_run1_df['avg_val_rmse'].std()
CNN_FCN_run1_std_rmse

0.13234056729733196

In [51]:
(CNN_FCN_run1_std_rmse * EE_STD) / 1000

0.3642276197211529

## Run 2 80/20 Split

In [23]:
CNNFCN_run2_path = 'results/CNNFCN_run2.csv'
CNNFCN_run2_df = pd.read_csv(CNNFCN_run2_path)
CNNFCN_run2_df

train_mse   val_mse    val_r2  \
0   0.479469  0.338891  0.719403   

                                          model_name  epoch  
0  CNNFCNRegressor_conv3_fcn3_fcnw32_lr0.01_batch...     38

In [71]:
CNN_RMSE_run2 = np.sqrt(CNNFCN_run2_df['val_mse'].values[0])
CNN_RMSE_run2 = (CNN_RMSE_run2 * EE_STD + EE_MEAN) / 1000
CNN_RMSE_run2

5.083801329918369

In [72]:
CNN_RMSE_run2 / WEIGHT_MEAN / 1000

6.938219307010718e-05

CNN 80/20 RMSE: 

5.083 kcal/min

0.0694 kcal/kg/min

In [73]:
CNN_CV_run2 = CNN_RMSE_run2 / (EE_MEAN / 1000) # kcal/min
CNN_CV_run2

1.4601804545434378

# CNNLSTMFCN

## Run 1 LOOCV

### Best model

In [17]:
CNNLSTMFCN_run1_path = 'results/CNNLSTMFCN_run1.csv'
CNNLSTMFCN_run1_df = pd.read_csv(CNNLSTMFCN_run1_path)
# CNNLSTMFCN_run1_df

In [13]:
CNNLSTMFCN_run1_min_rmse = CNNLSTMFCN_run1_df[CNNLSTMFCN_run1_df['avg_val_rmse'] == CNNLSTMFCN_run1_df['avg_val_rmse'].min()]
CNNLSTMFCN_run1_min_rmse

avg_train_mse  avg_train_rmse  avg_val_mse  avg_val_rmse  avg_val_r2  \
15       0.369113        0.607547     0.535163      0.731548    0.479545   

    avg_epochs                                         model_name   timestamp  \
15   38.272727  CNNLSTMFCNRegressor_conv3_fcn3_fcnw32_lstm1_ls...  1732813474   

    checkpoint_dir_name  done  ...  config/fcn_width config/lstm_hidden_size  \
15                  NaN  True  ...                32                      64   

    config/lstm_num_layers                                     config/log_dir  \
15                       1  /Users/tyler/GitHub Repositories/UT Austin/202...   

                                     config/model_dir  \
15  /Users/tyler/GitHub Repositories/UT Austin/202...   

   config/early_stop_patience config/fold_early_stop  config/print  \
15                         20                     20         False   

    config/tune config/loocv  
15         True         True  

[1 rows x 36 columns]

In [ ]:
CNNLSTMFCN_run1_min_rmse['model_name'].values[0]

'CNNLSTMFCNRegressor_conv3_fcn3_fcnw32_lstm1_lstmh64_lr0.01_batch32_epochs100'

In [15]:
CNNLSTMFCN_run1_max_r2 = CNNLSTMFCN_run1_df[CNNLSTMFCN_run1_df['avg_val_r2'] == CNNLSTMFCN_run1_df['avg_val_r2'].max()]
CNNLSTMFCN_run1_max_r2

avg_train_mse  avg_train_rmse  avg_val_mse  avg_val_rmse  avg_val_r2  \
15       0.369113        0.607547     0.535163      0.731548    0.479545   

    avg_epochs                                         model_name   timestamp  \
15   38.272727  CNNLSTMFCNRegressor_conv3_fcn3_fcnw32_lstm1_ls...  1732813474   

    checkpoint_dir_name  done  ...  config/fcn_width config/lstm_hidden_size  \
15                  NaN  True  ...                32                      64   

    config/lstm_num_layers                                     config/log_dir  \
15                       1  /Users/tyler/GitHub Repositories/UT Austin/202...   

                                     config/model_dir  \
15  /Users/tyler/GitHub Repositories/UT Austin/202...   

   config/early_stop_patience config/fold_early_stop  config/print  \
15                         20                     20         False   

    config/tune config/loocv  
15         True         True  

[1 rows x 36 columns]

In [16]:
CNNLSTMFCN_run1_max_r2['model_name'].values[0]

'CNNLSTMFCNRegressor_conv3_fcn3_fcnw32_lstm1_lstmh64_lr0.01_batch32_epochs100'

In [ ]:
CNNLSTM_RMSE = CNNLSTMFCN_run1_min_rmse['avg_val_rmse'].values[0]
CNNLSTM_RMSE = (CNNLSTM_RMSE * EE_STD + EE_MEAN) / 1000 # kcal/min
CNNLSTM_RMSE

5.494991824533308

In [57]:
CNNLSTM_RMSE / WEIGHT_MEAN

0.07499399739418068

CNN-LSTM RMSE: 

5.494 kcal/min

0.0750 kcal/kg/min

In [59]:
CNNLSTM_CV = CNNLSTM_RMSE / (EE_MEAN / 1000) # kcal/min
CNNLSTM_CV

1.578283481071507

### All models

In [52]:
CNNLSTMFCN_run1_avg_rmse = CNNLSTMFCN_run1_df['avg_val_rmse'].mean()
CNNLSTMFCN_run1_avg_rmse

0.8956828881781534

In [53]:
(CNNLSTMFCN_run1_avg_rmse * EE_STD + EE_MEAN) / 1000

5.946723218466599

In [54]:
CNNLSTM_FCN_run1_std_rmse = CNNLSTMFCN_run1_df['avg_val_rmse'].std()
CNNLSTM_FCN_run1_std_rmse

0.12448309382514211

In [55]:
(CNNLSTM_FCN_run1_std_rmse * EE_STD) / 1000

0.34260228655050157

## Run 2 80/20 Split

In [35]:
CNNLSTMFCN_run2_path = 'results/CNNLSTMFCN_run2.csv'
CNNLSTMFCN_run2_df = pd.read_csv(CNNLSTMFCN_run2_path)
CNNLSTMFCN_run2_df

train_mse   val_mse    val_r2  \
0   0.438891  0.341947  0.716369   

                                          model_name  epoch  
0  CNNLSTMFCNRegressor_conv3_fcn3_fcnw32_lstm1_ls...     43

In [38]:
CNNLSTM_RMSE_run2 = np.sqrt(CNNLSTMFCN_run2_df['val_mse'].values[0])
CNNLSTM_RMSE_run2 = CNNLSTM_RMSE_run2 * EE_STD + EE_MEAN
CNNLSTM_RMSE_run2

5091.008998352013

In [43]:
CNNLSTM_RMSE_run2 / WEIGHT_MEAN / 1000

0.06948056116327112

CNN 80/20 RMSE: 

5.091 kcal/min

0.0695 kcal/kg/min